src/internal/m365/collection/exchange/container_resolver.go

```diff
@@ -415,3 +416,125 @@ func (cr *containerResolver) populatePaths(
 
 	return lastErr
 }
+
+// ---------------------------------------------------------------------------
+// rankedContainerResolver
+// ---------------------------------------------------------------------------
+
+type rankedContainerResolver struct {
+	graph.ContainerResolver
+	// resolvedIncluded is the ordered list of actual container IDs to add to the
+	// start of the Items result set.
+	resolvedInclude []string
+	// resolvedExclude is the set of items that shouldn't be included in the
+	// result of Items or ItemByID. Uses actual container IDs instead of
+	// well-known names.
+	resolvedExclude map[string]struct{}
+}
+
+// newRankedContainerResolver creates a wrapper around base that returns results
+// from Items in priority order. Priority is defined by includeRankedIDs. All
+// items that don't appear in includeRankedIDs are considered to have equal
+// priority but lower priority than those in includeRankedIDs.
+//
+// includeRankedIDs is the set of containers to place at the start of the result
+// of Items in the order they should appear. IDs can either be actual
+// container IDs or well-known container IDs like "inbox".
+//
+// excludeIDs is the set of IDs that shouldn't be in the results returned by
+// Items. IDs can either be actual container IDs or well-known container IDs
+// like "inbox".
+func newRankedContainerResolver(
+	ctx context.Context,
+	base graph.ContainerResolver,
+	getter containerGetter,
+	userID string,
+	includeRankedIDs []string,
+	excludeIDs []string,
+) (*rankedContainerResolver, error) {
+	if base == nil {
+		return nil, clues.New("nil base ContainerResolver")
+	}
+
+	cr := &rankedContainerResolver{
+		resolvedInclude:   make([]string, 0, len(includeRankedIDs)),
+		resolvedExclude:   make(map[string]struct{}, len(excludeIDs)),
+		ContainerResolver: base,
+	}
+
+	for _, id := range includeRankedIDs {
+		ictx := clues.Add(ctx, "container_id", id)
+
+		c, err := getter.GetContainerByID(ctx, userID, id)
+		if err != nil {
+			return nil, clues.Wrap(err, "getting ranked container").WithClues(ictx)
+		}
+
+		gotID := ptr.Val(c.GetId())
+		if len(gotID) == 0 {
+			return nil, clues.New("received ranked include container with empty ID").
```

Comment:
Curious, does this need to be a failure? Am I correct in assuming that if we return a failure here, it would result in a preview backup failure?

Reply:
It's a good safety measure.  While graph api gives us good guarantees about the presence of an id, suppose we one day replace the inner item with an interface-compliant internal struct, or accidentally omit the ID from the item properties by selection.  This ensures we'll, at minimum, fail some integration tests.

Reply:
nit: the error here is a little verbose
```suggestion
			return nil, clues.New("ranked include container missing ID").
```

Reply:
> This ensures we'll, at minimum, fail some integration tests.

Understood. My concern is around user facing errors, especially for preview backups. Wondering if it's better to show user some less data(say one less folder) vs showing them a failed backup. We can always catch errors using integ tests but treat them as ignorable errors in prod.

I understand failures would be mostly unlikely here, but wanted to have a UX discussion. Doesn't need to be addressed in this PR.